### PR TITLE
created onParticleRequire strategy

### PR DESCRIPTION
**Pull request by AliceLetitia**
included require section in recipe clone

### CODE REVIEW

**Comment by shans** [comment]
This looks good, and is well tested.

Just to confirm: slots are matched at the moment but not handles yet. Is that correct? If so, it'd be good to add a TODO that notes this to the top of the strategy.

File: src/planning/test/strategies/find-required-particle-test.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
```
Comment (mykmartin) [inline src/planning/test/strategies/find-required-particle-test.ts]: Ditto

File: src/planning/strategies/find-required-particle.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2018 Google Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/planning/strategies/find-required-particle.ts]: Done :)

File: src/runtime/recipe/particle.ts [hunk 3]
```diff
@@ -99,6 +120,31 @@ export class Particle {
     return 0;
   }
 
+  /**
+   * Param particle matches this particle if the names are the same and the slot and handle requirements this particle 
+   * is a subset of the slot and handle requirements of the param particle. 
+   * @param particle 
+   */
+  matches(particle: Particle): boolean {
+    if (this.name && particle.name && !(this.name === particle.name)) return false;
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/runtime/recipe/particle.ts]: Done :)

File: src/runtime/recipe/particle.ts [hunk 4]
```diff
@@ -99,6 +120,31 @@ export class Particle {
...
+      
+      if (slotConn.targetSlot && slotConn.targetSlot.id && slotConn.targetSlot.id !== particle.consumedSlotConnections[name].targetSlot.id) return false;
+      
+      if ((Object.keys(slotConn.providedSlots)).some(pname => {
+        const slot = slotConn.providedSlots[pname];
+        const pslot = particle.consumedSlotConnections[name].providedSlots[pname];
+        if (pslot == undefined) return true;
+        if (slot.id && pslot.id && slot.id !== pslot.id) return true;
+        return false;
+      })) return false;
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/runtime/recipe/particle.ts]: Done :)

File: src/runtime/recipe/recipe.ts [hunk 5]
```diff
@@ -623,11 +646,32 @@ export class Recipe {
...
   }
 
   getDisconnectedConnections() {
     return this.handleConnections.filter(
         hc => hc.handle == null && !hc.isOptional && hc.name !== 'descriptions' && hc.direction !== 'host');
   }
 }
+
+export class RequireSection extends Recipe {
+  private _recipe: Recipe;
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/runtime/recipe/recipe.ts]: Done :)

File: src/runtime/recipe/recipe.ts [hunk 6]
```diff
@@ -623,11 +646,32 @@ export class Recipe {
...
+  constructor(recipe = undefined, name = undefined) {
+    super(name);
+    this.recipe = recipe;
+  }
+  set recipe(recipe) {this._recipe = recipe;}
+  get recipe() {return this._recipe; }
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/runtime/recipe/recipe.ts]: I agree. Done :)

File: src/runtime/recipe/slot-utils.ts [hunk 7]
```diff
@@ -5,31 +5,41 @@
...
+  static getClonedSlot(recipe, selectedSlot) {
+    let clonedSlot = recipe.updateToClone({selectedSlot}).selectedSlot;
+    if (!clonedSlot) {
+      if (selectedSlot.id) {
+        clonedSlot = recipe.findSlotByID(selectedSlot.id);
+      }
+      if (clonedSlot == undefined) {
+        if (recipe instanceof RequireSection) {
+          clonedSlot = recipe.recipe.newSlot(selectedSlot.name);
+          // recipe.addSlot(clonedSlot);
```
[the remark this answers is not in the document]
Comment (AliceLetitia) [reply src/runtime/recipe/slot-utils.ts]: (oops) Done :)

File: src/planning/test/strategies/find-required-particle-test.ts (hunk 1 above)
Comment (AliceLetitia) [reply src/planning/test/strategies/find-required-particle-test.ts]: Done :)

**Comment by AliceLetitia** [comment]
I've added another test that makes sure the recipe doesn't resolve when the strategizer finds a pattern in the active recipe that *almost* matches the require section.

**Pull request merged**